src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs

```diff
@@ -6369,5 +6369,678 @@ private void Handle3(SymbolAnalysisContext context)
+{
+  // Code size       19 (0x13)
+  .maxstack  2
+  .locals init (T V_0)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_0
+  IL_0004:  ldc.i4.s   42
+  IL_0006:  constrained. ""T""
+  IL_000c:  callvirt   ""void I.X.set""
+  IL_0011:  ldloc.0
+  IL_0012:  ret
+}");
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree, ignoreAccessibility: false);
+            var with = tree.GetRoot().DescendantNodes().OfType<WithExpressionSyntax>().Single();
+            var type = model.GetTypeInfo(with);
+            Assert.Equal("T", type.Type.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void WithExprOnStruct_OnRecordStructParameter()
+        {
+            var src = @"
+var b = new B(1);
+var b2 = B.M(b);
+System.Console.Write(b2.X);
+
+public record struct B(int X)
+{
+    public static B M(B b)
+    {
+        return b with { X = 42 };
+    }
+}";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "42");
+            verifier.VerifyIL("B.M", @"
+{
+  // Code size       13 (0xd)
+  .maxstack  2
+  .locals init (B V_0)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_0
+  IL_0004:  ldc.i4.s   42
+  IL_0006:  call       ""void B.X.set""
+  IL_000b:  ldloc.0
+  IL_000c:  ret
+}");
+        }
+
+        [Fact]
+        public void WithExprOnStruct_OnRecordStructParameter_Readonly()
+        {
+            var src = @"
+var b = new B(1, 2);
+var b2 = B.M(b);
+System.Console.Write(b2.X);
+System.Console.Write(b2.Y);
+
+public readonly record struct B(int X, int Y)
+{
+    public static B M(B b)
+    {
+        return b with { X = 42, Y = 43 };
+    }
+}";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "4243", verify: Verification.Skipped /* init-only */);
+            verifier.VerifyIL("B.M", @"
+{
+  // Code size       22 (0x16)
+  .maxstack  2
+  .locals init (B V_0)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_0
+  IL_0004:  ldc.i4.s   42
+  IL_0006:  call       ""void B.X.init""
+  IL_000b:  ldloca.s   V_0
+  IL_000d:  ldc.i4.s   43
+  IL_000f:  call       ""void B.Y.init""
+  IL_0014:  ldloc.0
+  IL_0015:  ret
+}");
+        }
+
+        [Fact]
+        public void WithExprOnStruct_OnTuple()
+        {
+            var src = @"
+var b = (1, 2);
+var b2 = M(b);
+System.Console.Write(b2.Item1);
+System.Console.Write(b2.Item2);
+
+static (int, int) M((int, int) b)
+{
+    return b with { Item1 = 42, Item2 = 43 };
+}";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "4243");
+            verifier.VerifyIL("<Program>$.<<Main>$>g__M|0_0(System.ValueTuple<int, int>)", @"
```

Comment:
><Program>$.<<Main>$>g__M|0_0(System.ValueTuple<int, int>) [](start = 31, length = 57)

It feels like it would be better to have a regualr method within a type.